src/lib/pubkey/pk_options.h

```diff
@@ -0,0 +1,143 @@
+      /// If prehash is nullopt, this requests prehashing using an algorithm
+      /// specific default function
+      ///
+      /// If the scheme does not support prehashing, it will throw an
+      /// exception when presented with such an option.
+      PK_Signature_Options with_prehash(std::optional<std::string> prehash = std::nullopt) const;
+
+      /// Specify a context
+      ///
+      /// Some signature schemes allow specifying a context with the signature.
+      /// This is typically a fixed string that identifies a protocol or peer.
+      ///
+      /// For SM2 this context is the user identifier
+      ///
+      /// If the scheme does not support contextual identifiers, then an exception
+      /// will be thrown.
+      PK_Signature_Options with_context(std::span<const uint8_t> context) const;
+
+      /// Specify a context as a string
+      ///
+      /// Equivalent to the version taking a span above; just uses the bytes
+      /// of the string instead.
+      PK_Signature_Options with_context(std::string_view context) const;
+
+      /// Request producing a deterministic signature
+      ///
+      /// Some signature schemes are always randomized, while others support a
+      /// randomized or deterministic option. This allows requesting this. For
+      /// signatures which are always deterministic, this option has no effect.
+      ///
+      /// This option is ignored for verification
+      PK_Signature_Options with_deterministic_signature() const;
+
+      /// Specify producing or expecting a DER encoded signature
+      ///
+      /// This is mostly used with ECDSA
+      ///
+      /// For schemes that do not support such formatting (such as RSA
+      /// or post-quantum schemes), an exception will be thrown when the
+      /// PK_Signer or PK_Verifier is created.
+      PK_Signature_Options with_der_encoded_signature() const;
+
+      /// Specify a provider that should be used
+      ///
+      /// This is rarely relevant
+      PK_Signature_Options with_provider(std::string_view provider) const;
+
+      const std::string& hash_function() const { return m_hash_fn; }
+
+      const std::optional<std::string>& prehash_fn() const { return m_prehash; }
+
+      const std::optional<std::string>& padding() const { return m_padding; }
+
+      const std::optional<std::vector<uint8_t>>& context() const { return m_context; }
+
+      const std::optional<std::string>& provider() const { return m_provider; }
+
+      bool using_der_encoded_signature() const { return m_use_der; }
+
+      bool using_deterministic_signature() const { return m_deterministic_sig; }
+
+      bool using_context() const { return context().has_value(); }
+
+      bool using_prehash() const { return m_use_prehash; }
+
+      bool using_padding() const { return padding().has_value(); }
+
+      bool using_provider() const { return provider().has_value() && provider().value() != "base"; }
+
+      // TODO
+      bool _valid_for_rsa() const;
```

Comment:
These are just examining the data and don't need access to the internals; probably should be moved to an internal header.